Change on_obj Function + Upgrade Vitessce

ilan-gold opened the pull request:
In this PR, I get rid of `on_obj` call back for `to_dict` on the view config object and internalize it by creating a new method on the configuration class, `get_routes`, to handle the behavior that `on_obj` seems to be used for repeatedly.

In the portal, we want to be able to call `to_dict` without any arguments and get the view config with the URLs from the wrappers/files.  In the current setup `to_dict` returns a view config with empty `files` field because it has no way of constructing the URLs with the `on_obj` callback.  The only catch with this method is that for local data, if you do not pass in a `base_url`, you could get back out a view config with completely unusable urls (as opposed to before, where `files` would just be empty).

keller-mark commented:
In the widget.py when `.get_routes` and `.to_dict` are both called, the `._get_data` functions will now be called twice (first for routes and then for the file definitions), and therefore creating the widget will be 2x as slow (I am thinking this will be an issue for the slow genomic data processing in particular). This is not necessarily a big deal, I just want to think about how we might be able to resolve that issue, perhaps by caching the processed data when the function runs the first time.

I am not sure I fully understand the use case in the portal. Are you planning to read in the AnnData zarr store with the `AnnDataWrapper(adata_url)` on-the-fly when the user loads a dataset in the portal?

keller-mark commented:
When I created the `on_obj` function I never really intended for it to be used with remote data. I assumed remote files would always be added with `add_file` instead of `add_object`. I see now that `add_object` is very helpful for both cases because the logic for the `"options"` field is typically the same for remote and local files.

How about if we introduce a separate type of wrapper class which is denoted as "remote" by inheriting from an abstract `RemoteWrapperType` ancestor class? This would allow us to flag those wrapper objects so that we know that we can ignore all "routes" details. Instead we can obtain the file definitions at the time of the `add_object` call (in other words, no need to wait until `to_dict` is called to generate the file URLs).

I tried out this approach on this branch: 
https://github.com/vitessce/vitessce-python/compare/keller-mark/remote_wrapper_classes

with example notebook here https://github.com/vitessce/vitessce-python/blob/keller-mark/remote_wrapper_classes/docs/notebooks/widget_imaging.ipynb

This resolves the issue of `._get_data` being called twice but does have the downside that the wrapper class imports are  more confusing.
- Perhaps instead of the separate classes with `Remote___` prefixes, one of the following would be simpler:
  - a `remote=True` parameter for all wrapper class constructors, or
  - inferring remote-ness based on the use of `img_url` vs `img_path`

keller-mark commented:
I thought about this more and I like the separate `get_routes` (so ignore my previous comment). I think maybe a better way to solve the `._get_data` issue is to avoid putting any data conversion steps in those functions. And then de-couple the creation of routes from the creation of data responses. 

This could also allow us to get rid of the `export_to_files` and `export_to_s3` functions, for example if all of the current temporary JSON responses are replaced by always saving a new JSON file + serving as static files. And it may be more clear to users that data conversion is happening if they are able to specify a directory for saving processed data.

ilan-gold commented:
@keller-mark That is slick, wow! I think it really gives a nicer semantic to our whole data model. I can take the hand off from you on that branch and polish it up (if need be)/fix docs if that works.  I'll then re-request your review to give it a final look.

keller-mark commented:
Ok thank you! I am not sure why I introduced the complexity of the specific `get_raster`, `get_cells`, etc. in the first place, when ultimately we just need the server routes and the file definitions. I will focus on polishing up the demo data processing branch

ilan-gold commented:
I need to look this over again in the morning but I seem to have set up all the data wrappers properly.  So far here's the change list:  
- I removed the `OmeZarrWrapper` because it's so far out of sync at this point and I think we need to revisit the whole zarr side of things anyway for imaging. 
- I needed to keep the `JSONRoute` for the `offsets.json` to avoid having a clash of routes for multi-modal data (i.e multiple `OmeTiffWrapper` objects under a `MultiImageWrapper` will have the same `dataset_uid` and `obj_i` so they will try mounting the directories that contain the `offsets.json` twice but with different directories).
- I also added `is_remote` to the base class.
- Create a `get_route` method to the parent class of some sort for just serving the `out_dir`.  I need to rename this to make it clear exactly what this method does.

keller-mark reviewed (approved):
Sorry for the delay on this. I made a small PR to remove `create_obj_routes` and `create_obj_files` which do not appear to be used anymore, and refactored the `from_object` function so that the logic for the "auto view configuration" is in the wrapper classes https://github.com/vitessce/vitessce-python/pull/47 I tested by running the `widget_` notebooks and everything was working except `from_object` so looks good!
